### PR TITLE
[Task #1916310] Support customizing port for Azure Functions Core tools when run Function project

### PR DESCRIPTION
**Pull request by Flanker32**
### Issue to resolve
Currently toolkit will detect free port for function core tools automatically during run functions locally, however, this did not works well in mac for two issues:
- Azure Functions local process not killed properly on Mac
- Logic to determine port is free did not work correctly

### What this PR do
- Support specify port for function core tools in function run configuration
- Migrate to use `OSProcessUtil` to kill `func` and its child process
- Fix logic to determine port, refers https://stackoverflow.com/questions/434718/sockets-discover-port-availability-using-java
- Add action to retry with free port when meet port in use issue during run function locally

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-appservice/src/main/java/com/microsoft/azure/toolkit/intellij/legacy/function/runner/core/FunctionUtils.java [hunk 1]
```diff
@@ -544,4 +547,25 @@ private static boolean isModuleInTestScope(Module module) {
...
+                // swallow this exception
+            } finally {
+                if (socket != null) {
+                    try {
+                        socket.close();
+                    } catch (IOException e) {
+                        // swallow this exception
+                    }
+                }
+            }
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-appservice/src/main/java/com/microsoft/azure/toolkit/intellij/legacy/function/runner/core/FunctionUtils.java:564]: the `try` block may be simplified with `try resource`.

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-appservice/src/main/java/com/microsoft/azure/toolkit/intellij/legacy/function/runner/core/FunctionUtils.java [hunk 2]
```diff
@@ -544,4 +547,25 @@ private static boolean isModuleInTestScope(Module module) {
         }
         return cme.getCompilerOutputUrl() == null && cme.getCompilerOutputUrlForTests() != null;
     }
+
+    public static int findFreePortForApi(int startPort) {
+        ServerSocket socket = null;
+        for (int port = startPort; port <= MAX_PORT; port++) {
+            try {
+                socket = new ServerSocket(port);
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-appservice/src/main/java/com/microsoft/azure/toolkit/intellij/legacy/function/runner/core/FunctionUtils.java]: > We can use a special port number zero in the `ServerSocket` class constructor. As a result, the java.net API will automatically allocate a free port for us

can you try with `zero` port to let `ServerSocket` to find a free port for us?

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-appservice/src/main/java/com/microsoft/azure/toolkit/intellij/legacy/function/runner/localrun/FunctionRunConfiguration.java [hunk 3]
```diff
@@ -217,6 +225,9 @@ public void validate() throws ConfigurationException {
         if (!func.exists() || !func.isFile() || !func.getName().contains("func")) {
             throw new ConfigurationException(message("function.run.validate.invalidFuncPath"));
         }
+        if (getFuncPort() <= 0 || getFuncPort() >= 65535) {
+            throw new ConfigurationException(message("function.validate_run_configuration.invalidPort"));
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-appservice/src/main/java/com/microsoft/azure/toolkit/intellij/legacy/function/runner/localrun/FunctionRunConfiguration.java:229]: use `AzureToolkitRuntimeException` please.

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-appservice/src/main/java/com/microsoft/azure/toolkit/intellij/legacy/function/runner/localrun/FunctionRunModel.java [hunk 4]
```diff
@@ -65,4 +66,11 @@ public void setLocalSettingsJsonPath(String localSettingsJsonPath) {
         this.localSettingsJsonPath = localSettingsJsonPath;
     }
 
+    public int getFuncPort() {
+        return funcPort;
+    }
+
+    public void setFuncPort(int funcPort) {
+        this.funcPort = funcPort;
+    }
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-appservice/src/main/java/com/microsoft/azure/toolkit/intellij/legacy/function/runner/localrun/FunctionRunModel.java]: why not use `@Getter`/`@Setter`

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-appservice/src/main/java/com/microsoft/azure/toolkit/intellij/legacy/function/runner/core/FunctionUtils.java [hunk 5]
```diff
@@ -568,4 +568,14 @@ public static int findFreePortForApi(int startPort) {
...
     }
+
+    public static int getFreePort() {
+        try {
+            ServerSocket serverSocket = new ServerSocket(0);
+            serverSocket.close();
+            return serverSocket.getLocalPort();
+        } catch (IOException e) {
+            return -1;
+        }
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-appservice/src/main/java/com/microsoft/azure/toolkit/intellij/legacy/function/runner/core/FunctionUtils.java:574]: can be simplified with `try resource`

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-appservice/src/main/java/com/microsoft/azure/toolkit/intellij/legacy/function/runner/localrun/FunctionRunConfiguration.java (hunk 3 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-appservice/src/main/java/com/microsoft/azure/toolkit/intellij/legacy/function/runner/localrun/FunctionRunConfiguration.java:229]: This is to follow the framework of https://github.com/microsoft/azure-tools-for-java/blob/develop/PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-appservice/src/main/java/com/microsoft/azure/toolkit/intellij/legacy/common/AzureRunConfigurationBase.java#L43, which will be called in `SettingsEditor.applyEditorTo`, uses `ConfigurationException`.

**Review by wangmingliang-ms — APPROVED** [review]

**Pull request merged**